### DISCUSSION TITLE
Persisting/retrieving state - Improvements

### DISCUSSION THREAD

**Discussion by RemyMachado** [statelyai/xstate#3654] [Ideas]
I just started using xstate and I must say that it's a great tool!

However, I feel like some mechanics of the lib' are not straight forward.
First thing I needed was storing the state of my machine into my DB, then restore it later.

First issue I encountered was "How to retrieve the state at any given time?". It was `service.getSnapshot()`, and I didn't see it in the whole `guides` section of the documentation (was only informed of `newState = service.send()`).

Second issue, after following the [persisting state doc section](https://xstate.js.org/docs/guides/states.html#persisting-state), I couldn't test the `.can()` of the restored state.
```typescript
const stateDefinition =
  JSON.parse(localStorage.getItem('app-state')) || myMachine.initialState;

// Use State.create() to restore state from a plain object
const previousState = State.create(stateDefinition);

previousState.can({type: 'type'}); // always false
```

I even tried interpreting it back.

Little did I know, that we have to resolve the state.
```typescript
const resolvedState = machine.resolveState(state);
``` 

The trouble was too severe, for something that simple.
It may be obvious when you truly understand how the library works, but from a beginner point of view I was quite confused.

### I'm sharing this feedback in order to maybe improve the documentation or the library. I don't know, what do you think?

**Comment by davidkpiano** [comment]
You're right; we should improve the documentation here, especially since `.getSnapshot()` and `.can()` are relatively recent additions to the API.
